packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -272,6 +272,46 @@ abstract mixin class WidgetsBindingObserver {
 }
 
 /// The glue between the widgets layer and the Flutter engine.
+///
+/// The [WidgetsBinding] manages a single [Element] tree rooted at [rootElement].
+/// Calling [runApp] (which indirectly calls [attachRootWidget]) bootstraps that
+/// element tree.
+///
+/// ## Relationship to render trees
+///
+/// Multiple render trees may be associated with the element tree. Those are
+/// managed by the underlying [RendererBinding].
+///
+/// The element tree is segmented into two types of zones: rendering zones and
+/// non-rendering zones.
+///
+/// A rendering zone is a part of the element tree that is backed by a render
+/// tree and it describes the pixels that are drawn on screen. For elements in
+/// this zone [Element.renderObject] never returns null because the elements are
+/// all associated with [RenderObject]s. Almost all widgets can be placed in a
+/// rendering zone, with the [View] widget, [ViewCollection] widget, and
+/// [RootWidget] as notable exceptions.
+///
+/// A non-rendering zone is a part of the element tree that is not backed by a
+/// render tree. For elements in this zone [Element.renderObject] returns null
+/// because the elements are not associated with any [RenderObject]s. Only
+/// widgets that do not produce a [RenderObject] can be used in this zone
+/// because there is no render tree to attach the render object to. In other
+/// words, [RenderObjectWidget]s cannot be used in this zone. Typically, one
+/// would find [InheritedWidget]s, [View]s, and [ViewCollection]s in this zone
+/// to inject data across rendering zones into the tree and to organize the
+/// rendering zones (and by extension their associated render trees) into a
+/// unified element tree.
+///
+/// The root of the element tree at [rootElement] defines the beginning of a
+/// non-rendering zone. The [View] widget is used to start a rendering zone
+/// because it bootstraps a render tree. Within a rendering zone, the
+/// [ViewAnchor] may be used to create a new non-rendering zone.
+///
```

Comment:
At some point including an example graph here would be nice, but it would have to live in https://github.com/flutter/assets-for-api-docs/ and therefore be generated from a flutter app. I am going to leave that work for a later PR (added a TODO).